choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java

```diff
@@ -0,0 +1,84 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.ballerina.observe.choreo;
+
+import io.ballerina.observe.choreo.logging.LogFactory;
+import io.ballerina.observe.choreo.logging.Logger;
+import io.ballerina.runtime.observability.BallerinaObserver;
+import io.ballerina.runtime.observability.ObserverContext;
+import io.ballerina.runtime.observability.metrics.DefaultMetricRegistry;
+import io.ballerina.runtime.observability.metrics.MetricId;
+import io.ballerina.runtime.observability.metrics.MetricRegistry;
+import io.ballerina.runtime.observability.metrics.Tag;
+
+import java.util.Set;
+
+/**
+ * Observe the runtime and collect measurements.
+ */
+
+public class StepCountObserver implements BallerinaObserver {
+    private static final String PROPERTY_START_TIME = "_choreo_observation_start_time_";
+
+    private static final String PROPERTY_STEP_TIME = System.getenv("choreo_per_step_time");
+         
+    private static final Logger LOGGER = LogFactory.getLogger();
+
+    @Override
+    public void startServerObservation(ObserverContext observerContext) {
+        startObservation(observerContext);
+    }
+    @Override
+    public void startClientObservation(ObserverContext observerContext) {
+        startObservation(observerContext);
+    }
+    @Override
+    public void stopServerObservation(ObserverContext observerContext) {
+        if (!observerContext.isStarted()) {
+            // Do not collect metrics if the observation hasn't started
+            return;
+        }
+        stopObservation(observerContext);
+    }
+    @Override
+    public void stopClientObservation(ObserverContext observerContext) {
+        if (!observerContext.isStarted()) {
+            // Do not collect metrics if the observation hasn't started
+            return;
+        }
+        stopObservation(observerContext);
+    }
+    
+    private void startObservation(ObserverContext observerContext) {
+        observerContext.addProperty(PROPERTY_START_TIME, System.currentTimeMillis());
+    }
+    
+    private void stopObservation(ObserverContext observerContext) {
+        Set<Tag> tags = observerContext.getAllTags();
+        try {
+            Long startTime = (Long) observerContext.getProperty(PROPERTY_START_TIME);
+            long stepTime = Long.parseLong(PROPERTY_STEP_TIME);
+            long duration = System.currentTimeMillis() - startTime;
+            long steps = Math.round(Math.ceil((double) (duration - stepTime) / stepTime));
+            final MetricRegistry metricRegistry = DefaultMetricRegistry.getInstance();
+            metricRegistry.counter(new MetricId("choreo_steps_total", "Total no of steps", tags)).increment(steps);
+        } catch (RuntimeException e) {
+            LOGGER.error("error: error collecting metrics for multiple metrics with tags " 
+            + tags + ": " + e.getMessage());
```

Comment:
Shall we exactly say what failed here in the error message ?
Also the `error: ` prefix is automatically added by the logger and therefore we should remove it from here.
```suggestion
            LOGGER.error("failed to record choreo_steps_total metric with tags " 
                + tags + ": " + e.getMessage());
```